src/Nethermind/Nethermind.Evm/IL/ILVirtualMachineBuilder.cs

```diff
@@ -0,0 +1,426 @@
+        LocalBuilder consumeJumpCondition = il.DeclareLocal(typeof(int));
+        LocalBuilder uint256A = il.DeclareLocal(typeof(UInt256));
+        LocalBuilder uint256B = il.DeclareLocal(typeof(UInt256));
+        LocalBuilder uint256C = il.DeclareLocal(typeof(UInt256));
+
+        LocalBuilder gasAvailable = il.DeclareLocal(typeof(long));
+
+        // TODO: stack check for head
+        LocalBuilder stack = il.DeclareLocal(typeof(Word*));
+        LocalBuilder current = il.DeclareLocal(typeof(Word*));
+
+        const int wordToAlignTo = 32;
+
+        il.Emit(OpCodes.Ldc_I4, EvmStack.MaxStackSize * Word.Size + wordToAlignTo);
+        il.Emit(OpCodes.Localloc);
+
+        // align to the boundary, so that the Word can be written using the aligned longs.
+        il.LoadValue(wordToAlignTo);
+        il.Emit(OpCodes.Conv_I);
+        il.Emit(OpCodes.Add);
+        il.Emit(OpCodes.Ldc_I4, ~(wordToAlignTo - 1));
+        il.Emit(OpCodes.Conv_I);
+        il.Emit(OpCodes.And);
+
+        il.Store(stack); // store as start
+
+        il.Load(stack);
+        il.Store(current); // copy to the current
+
+        // gas
+        il.Emit(OpCodes.Ldarg_0);
+        il.Store(gasAvailable);
+        Label outOfGas = il.DefineLabel();
+
+        Label ret = il.DefineLabel(); // the label just before return
+        Label invalidAddress = il.DefineLabel(); // invalid jump address
+        Label jumpTable = il.DefineLabel(); // jump table
+
+        Dictionary<int, Label> jumpDestinations = new();
+
+        for (int pc = 0; pc < code.Length; pc++)
+        {
+            Operation op = Operation.Operations[(Instruction)code[pc]];
+
+            if (gasCost.TryGetValue(pc, out long required))
+            {
+                // check required > available
+                il.LoadValue(required);
+                il.Load(gasAvailable);
+                il.Emit(OpCodes.Cgt);
+                il.Emit(OpCodes.Brtrue, outOfGas); // jump to out of gas
+
+                // gasAvailable = gasAvailable - required
+                il.Load(gasAvailable);
+                il.LoadValue(required);
+                il.Emit(OpCodes.Sub);
+                il.Store(gasAvailable);
+            }
+
+            switch (op.Instruction)
+            {
+                case Instruction.PC:
+                    il.CleanWord(current);
+                    il.Load(current);
+                    il.LoadValue(BinaryPrimitives.ReverseEndianness(pc)); // TODO: assumes little endian machine
+                    il.Emit(OpCodes.Stfld, Word.UInt0Field);
+                    il.StackPush(current);
+                    break;
+
+                // pushes work as follows
+                // 1. load the next top pointer of the stack
+                // 2. zero it
+                // 3. load the value
+                // 4. set the field
+                // 5. advance pointer
+                case Instruction.PUSH1:
+                    il.Load(current);
```

Comment:
I could not really understand how stack works in msil. Could you expand this with more knowledges?
I feel like everything is operating on this `current` local.

Reply:
Yes, the `current` here is the top of the evm stack. If `PUSH1` is executed, the plan for it is flushed out above between lines 122-127. Load the value, zero it, set 1, advance.